src/main/java/com/minecolonies/client/gui/WindowHutFarmer.java

```diff
@@ -1,243 +1,316 @@
+            building.changeFields(field.getId(), true, row);
         }
-        building.melon++;
-        removeOthers(MELON);
-    }
 
+        pullLevelsFromHut();
+        window.findPaneOfTypeByID(LIST_FIELDS, ScrollingList.class).refreshElementPanes();
+    }
 
     /**
-     * Action performed when pumpkin button is clicked
-     *
-     * @param ignored   Parameter is ignored, since some actions require a button.
-     *                  This method does not
+     * Retrieve levels from the building to display in GUI
      */
-    private void pumpkinClicked(Button ignored)
+    private void pullLevelsFromHut()
     {
-        if (building.pumpkin >= MAX_AMOUNT)
-        {
-            return;
-        }
-        building.pumpkin++;
-        removeOthers(PUMPKIN);
+        fields = building.getFields();
     }
 
-
     /**
-     * Action performed when carrot button is clicked
+     * Fired when the assignment mode has been toggled.
      *
-     * @param ignored   Parameter is ignored, since some actions require a button.
-     *                  This method does not
+     * @param button clicked button.
      */
-    private void carrotClicked(Button ignored)
+    private void assignmentModeClicked(@NotNull Button button)
     {
-        if (building.carrot >= MAX_AMOUNT)
+        if (button.getLabel().equals(LanguageHandler.format("com.minecolonies.gui.hiring.off")))
         {
-            return;
+            button.setLabel(LanguageHandler.format("com.minecolonies.gui.hiring.on"));
+            building.setAssignFieldManually(true);
         }
-        building.carrot++;
-        removeOthers(CARROT);
+        else
+        {
+            button.setLabel(LanguageHandler.format("com.minecolonies.gui.hiring.off"));
+            building.setAssignFieldManually(false);
+        }
+        window.findPaneOfTypeByID(LIST_FIELDS, ScrollingList.class).refreshElementPanes();
     }
 
+    @NotNull
     @Override
     public String getBuildingName()
     {
-        return "com.minecolonies.gui.workerHuts.farmer";
+        return "tile.minecolonies.blockHutFarmer.name";
     }
 
     @Override
     public void onOpened()
     {
         super.onOpened();
 
-        updateButtonLabels();
+        if (building.assignFieldManually())
+        {
+            findPaneOfTypeByID(TAG_BUTTON_ASSIGNMENT_MODE, Button.class).setLabel(LanguageHandler.format("com.minecolonies.gui.hiring.on"));
+        }
+        else
+        {
+            findPaneOfTypeByID(TAG_BUTTON_ASSIGNMENT_MODE, Button.class).setLabel(LanguageHandler.format("com.minecolonies.gui.hiring.off"));
+        }
+
         findPaneOfTypeByID(BUTTON_PREV_PAGE, Button.class).setEnabled(false);
         buttonPrevPage = findPaneOfTypeByID(BUTTON_PREV_PAGE, Button.class);
         buttonNextPage = findPaneOfTypeByID(BUTTON_NEXT_PAGE, Button.class);
+
+        fieldList = findPaneOfTypeByID(LIST_FIELDS, ScrollingList.class);
+        fieldList.setDataProvider(new ScrollingList.DataProvider()
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this anonymous class number of lines from 52 to at most 30, or make it a named class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1188)